rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,723 @@
+   unnecessarily surprising and unsafe. No useful optimization opportunities are
+   opened by this undefined behavior, since volatile operations cannot be
+   meaningfully optimized.
+2. The absence of a hard guarantee that each volatile load or store will
+   translate into exactly one load or store at the hardware level is needlessly
+   tracherous in scenarios where memory access patterns must be very precisely
+   controlled, such as memory-mapped I/O.
+
+Both of these problems can be resolved by using atomic volatile operations
+instead of non-atomic volatile operations. However, in doing so, we must be very
+careful to stick with the subset of atomic operations that is supported by all
+Rust hardware targets.
+
+This means that we should not expose anything more than loads or stores in our
+standard subset, as some hardware architectures which are either very old or
+lack multiprocessing support do not support more than that.
+
+More surprisingly, maximizing hardware support also means that we cannot rely on
+`Ordering::Relaxed` as a minimal level of atomic sanity, because this ordering
+assumes a global order of operations targeting a single memory location across
+all threads in the program, and this property is not provided by native
+load/store operations on some targets without global cache coherence such as GPU
+architectures (NVPTX, AMD GCN, ...).
+
+As a result, we need to go deeper and expose an even weaker form of atomicity,
+which is more akin to LLVM's `unordered` atomic ordering semantics, but better
+specified in the context of atomic volatile operations because we provide very
+strong control on the sequence of emitted hardware memory accesses, and thus
+developers can refer to their hardware's memory model for precise semantics.
+
+One design goal of the proposed semantics is that it should be possible to
+implement `VolatileXyz::[load|store]` by compiling it down to LLVM unordered
+atomic volatile load/store operations of the same width, without specifying it
+in terms of "whatever LLVM unordered does" as that would make the life of
+alternate rustc backends like cranelift harder.
+
+## API design
+[api-design]: #api-design
+
+Switching to LLVM's atomic volatile accesses without also changing the API of
+`ptr::[read|write]_volatile` would unfortunately not resolve the memory access
+tearing problem very well, because although oversized volatile accesses would
+not compile anymore, this fact would only be "reported" to the user via an LLVM
+crash. This is not a nice user experience, which is why volatile wrapper types
+are proposed instead.
+
+Their design borrows from that of Rust's existing atomic types, which is only
+appropriate since they do expose atomic operations. The current proposal would
+be to fill the newly built `std::volatile` module with the following entities
+(some of which may not be available on a given platform, we will come back to
+this point in a moment):
+
+- `VolatileBool`
```

Comment:
New question: Do we actually want to have a `VolatileBool`, given that it has data invariants?

- Pro: Consistent with `AtomicBool` prior art.
- Con: Several core use cases for `VolatileXyz` involve interaction with entities that are not guaranteed to uphold Rust's invariants (memory-mapped hardware, untrusted code controlled by attacker...)
- Maybe-con: I'm not sure if Rust's `bool` is actually well-defined enough from an ABI PoV in order to be used for low-level hardware and software interfaces?

Reply:
I'd start without `VolatileBool` and add it when someone comes along with a reasonable use-case.

Reply:
Ok, I turned it into a future extension.